### DISCUSSION TITLE
How would I minimise change in portfolio weights while adhering to a risk limit constraint?

### DISCUSSION THREAD

**Discussion by NervousEnergy1979** [cvxpy/cvxpy#1817] [Q&A]
I have a set of portfolio target weights `w` which violate some constraints. I'm using cvxpy to find new weights as close as possible to these target weights, but that adhere to the constraints as in the code below.

```
x = cv.Variable(n)  # n = number of assets
prob = cv.Problem(
cv.Minimize(cv.sum(cv.abs(x - w))),  # w are my target weights
    [
        cv.abs(x) <= concentration_limit,
        cv.sum(x) == 0,  # Long/Short cash neutral portfolio
        cv.sum(cv.abs(x)) <= leverage_limit
    ]
)
prob.solve()
```

This works fine. However, say I wanted the portfolio annualised volatility within a limit, $[\sigma_{\text{min}}, \sigma_{\text{max}}]$. I see from [this example](https://colab.research.google.com/github/cvxpy/cvxpy/blob/master/examples/notebooks/WWW/SOC%20Portfolio.ipynb) that using SOC constraints I can achieve risk limits, but this is while maximising return, which I don't need for my problem.

How would I include risk constraints in the problem above? I'm sure it would involve using the covariance matrix of returns $\Sigma$, but I'm not sure how. I tried

```
cv.Minimize(cv.sum(cv.abs(x - w))),
    [
        cv.abs(x) <= concentration_limit,
        cv.sum(x) == 0,
        cv.sum(cv.abs(x)) <= leverage_limit,
        cv.quad_form(x, cov) * np.sqrt(252) * 100 <= sigma_max,  # cov = covariance matrix of returns
        cv.quad_form(x, cov) * np.sqrt(252) * 100 >= sigma_min,
    ]
)
```

But I get `DCPError: Problem does not follow DCP rules.`

**Comment by phschiele** [comment]
@ktattan The issue is that your last constraint is of the form `convex >= ...`
See the [DCP Rules](https://www.cvxpy.org/tutorial/dcp/index.html#dcp-problems) for a full overview of the allowed constraint types. 
In the portfolio context, one might argue that if one could find a closer portfolio according to your objective that _also_ has less risk, this should definitely be preferred.
